### DISCUSSION TITLE
Tasklist checkboxes inside a table

### DISCUSSION THREAD

**Discussion by NikosAlexandris** [facelessuser/pymdown-extensions#2305] [Q&A] [answered]
Dear all, is it possible to use the _Tasklist_ checkboxes inside a table ?

```
| What ? | To Do|
|--------------|-------------|
| - [ ] Task 1 | Something|
| - [x] Done | Nothing |
| [ ] Task 2 | |
| [x] Done 2| |
```
which gives : 

| What ? | To Do|
|--------------|-------------|
| - [ ] Task 1 | Something|
| - [x] Done | Nothing |
| [ ] Task 2 | |
| [x] Done 2| |

Or

```
|One|Two|
|--|--|
|- [ ]| More|
|- [x]| Even more?|
```
which gives

|One|Two|
|--|--|
|- [ ]| More|
|- [x]| Even more?|

and have it shown as outside the table, i.e.

- [ ] Something
- [x] Done

?

Or is the only option something like

```
| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | <ul><li>- [x] item1</li><li>- [ ] item2</li></ul>
| Object Cache   | > 5 hours  |  | in progress | <ul><li>[x] item1</li><li>[ ] item2</li></ul>
```

which shows as 

| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | <ul><li>- [x] item1</li><li>- [ ] item2</li></ul>
| Object Cache   | > 5 hours  |  | in progress | <ul><li>[x] item1</li><li>[ ] item2</li></ul>

?

Finally,
```
| Selection |        |
| --------- | ------ |
| <li>- [ ] </li> |  |
```
which is presented as 

| Selection |        |
| --------- | ------ |
| <li>- [x] </li> |  |

(last two examples taken from https://stackoverflow.com/a/47344640/1172302)

**Comment by facelessuser** [comment]
@NikosAlexandris Python Markdown won't parse Markdown in what it considers block HTML elements unless they are at the root level of the document and you use the `md_in_html` extension. Because these are embedded in tables, that makes it even more difficult for Python Markdown. So yes, list syntax is a block element in Python Markdown so it won't get parsed in a table. 

The short answer is `pymdownx.tasklist` will not work in the table syntax.

But fear not, there are ways to do what you want. I'm going highlight the easiest ways to do what you want. You can add classes and additional styling to manipulate the look.

## Raw HTML

If you absolutely want lists, and real HTML checkboxes, you can do this with raw HTML. You may also want to style them so the user can't toggle them, but I leave that up to you.

```md
### Raw HTML

| What ? | To Do|
|--------------|-------------|
| <input type="checkbox"> Task 1 | Something|
| <input type="checkbox" checked="checked"> Done | Nothing |
| <input type="checkbox"> Task 2 | |
| <input type="checkbox" checked="checked"> Done 2| |

| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | <ul><li><input type="checkbox" checked="checked"> item1</li><li><input type="checkbox"> item2</li></ul>
| Object Cache   | > 5 hours  |  | in progress | <ul><li><input type="checkbox" checked="checked"> item1</li><li><input type="checkbox"> item2</li></ul>
```

<img width="716" alt="Screenshot 2024-01-26 at 7 10 23 AM" src="https://github.com/facelessuser/pymdown-extensions/assets/1055125/3bfe5e70-0fd2-426f-b315-a589c02efe9b">


## Raw HTML and Emoji

If we don't explicitly need lists, let's cut down on the HTML tags and use Emoji using `pymdownx.emoji`. This is probably the easiest and simplest solution.

```md
### Emoji

| What ? | To Do|
|--------------|-------------|
| :green_square: Task 1 | Something|
| :white_check_mark: Done | Nothing |
| :green_square: Task 2 | |
| :white_check_mark: Done 2| |

| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | :white_check_mark: item1<br>:green_square: item2
| Object Cache   | > 5 hours  |  | in progress | :white_check_mark: item1<br>:green_square: item2
```

<img width="659" alt="Screenshot 2024-01-26 at 7 10 30 AM" src="https://github.com/facelessuser/pymdown-extensions/assets/1055125/37256d58-199e-4036-ae0a-5dc16728a95b">

## Combined Solution

You absolutely want lists, but like using the emoji? Okay, let's combine them.

```md
### Raw HTML and Emoji

| What ? | To Do|
|--------------|-------------|
| :green_square: Task 1 | Something|
| :white_check_mark: Done | Nothing |
| :green_square: Task 2 | |
| :white_check_mark: Done 2| |

| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | <ul><li>:white_check_mark: item1</li><li>:green_square: item2</li></ul>
| Object Cache   | > 5 hours  |  | in progress | <ul><li>:white_check_mark: item1</li><li>:green_square: item2</li></ul>
```

<img width="669" alt="Screenshot 2024-01-26 at 7 10 36 AM" src="https://github.com/facelessuser/pymdown-extensions/assets/1055125/1a7c52dd-993c-476e-82ff-e48f4cf34692">

**Reply by NikosAlexandris** [reply]
I was hoping for some programmatically more elaborated work and then less work of manual copy-pasting. Bothering is also that the emojis fully spelled-out, take quite some space in the case of wider tables :-/.

Anyway, thank you for your time and the examples! I have come quite some time ago, about, or exactly, to the same practice(s). Just a partial screenshot:
 
![2024-01-26T17:33:39,452808550+02:00](https://github.com/facelessuser/pymdown-extensions/assets/7046639/18681ce2-1e34-4d53-8825-fdf3fc6f5158)

However, what I mostly find useful in your answer is the "_You may also want to style them so the user can't toggle them, but I leave that up to you._". Specifically the first part "_You may also want to style them .._"!

**Can I style them to match the "default" Material for MkDocs look ?** I.e. to look exactly as the _Completed_ and _In Queue_ below ?

![2024-01-26T17:40:17,335376834+02:00](https://github.com/facelessuser/pymdown-extensions/assets/7046639/21162ff1-c141-4f99-aa16-f35e4e4d3d33)

Besides ease-of-use, I am after a _consequent_ look and feel in the documentation.

**Reply by facelessuser** [reply, marked as answer]
> I was hoping for some programmatically more elaborated work and then less work of manual copy-pasting.

Well, I unfortunately don't have any. My extensions work within the confines of Python Markdown, and I can't write custom extensions for everyone's cases.

> Can I style them to match the "default" Material for MkDocs look ? I.e. to look exactly as the Completed and In Queue below ?

Sure, for instance the checkmark icon that Material uses can be used via emoji with `:material-checkbox-marked-circle:` and the progress one is `:material-progress-question:`. You can search supported icons/emoji here: https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/?h=emoji.

You can choose how to target them with CSS, as a very simple example, I use `attr_list` extension to apply a class, though there are other approaches as well.

```md
### Emoji

| What ? | To Do|
|--------------|-------------|
| :material-checkbox-marked-circle:{.blank} Task 1 | Something|
| :material-checkbox-marked-circle:{.checked} Done | Nothing |
| :material-checkbox-marked-circle:{.blank} Task 2 | |
| :material-checkbox-marked-circle:{.checked} Done 2| |

| Task           | Time required | Assigned to   | Current Status | Finished | 
|----------------|---------------|---------------|----------------|-----------|
| Object Cache   | > 5 hours  |  | in progress | :material-checkbox-marked-circle:{.checked} item1<br>:material-checkbox-marked-circle:{.blank} item2
| Object Cache   | > 5 hours  |  | in progress | :material-checkbox-marked-circle:{.checked} item1<br>:material-checkbox-marked-circle:{.blank} item2

<style>
.twemoji.checked svg {
    color: #00e676;
}
.twemoji.blank svg {
    color: rgba(0, 0, 0, 0.07);
}
</style>
```

<img width="684" alt="Screenshot 2024-01-26 at 9 34 19 AM" src="https://github.com/facelessuser/pymdown-extensions/assets/1055125/1ce48cf8-fe7b-4d40-bbf1-7ef00d10172b">

That is probably all the time I can spend on this. I wish you luck though. If others want to chime in they can.

**Reply by NikosAlexandris** [reply]
This last _style_ bit is a great addition!  Thank you @facelessuser.
